File: abc425/e/main2.cpp
#include <bits/stdc++.h>
#include <atcoder/modint>
#include <atcoder/math>
using namespace std;
using ll = long long;
using ull = unsigned long long;
using mint = atcoder::modint998244353;
constexpr ll INF = 1e18;
template<typename T> inline void chmin(T &x, T y) { if (y < x) x = y; }
template<typename T> inline void chmax(T &x, T y) { if (x < y) x = y; }

int main() {
    ll t, m;
    cin >> t >> m;
    
    vector<bool> is_prime(5001, true);
    vector<ll> min_factor(5001, -1);
    is_prime[1] = false;
    min_factor[1] = 1;
    for (ll p = 2; p <= 5000; ++p) {
        if (!is_prime[p]) continue;
        min_factor[p] = p;
        for (ll q = 2*p; q <= 5000; q += p) {
            is_prime[q] = false;
            if (min_factor[q] == -1) min_factor[q] = p;
        }
    }

    // n!の素因数分解結果
    vector<unordered_map<ll, ll>> factorized(5001);
    for (ll x = 2; x <= 5000; ++x) {
        factorized[x] = factorized[x-1];
        ll y = x;
        while (y>1) {
            ll p = min_factor[y];
            while (min_factor[y] == p) {
                y /= p;
                ++factorized[x][p];
            }
        }
    }

    while (t--) {
        ll n;
        cin >> n;
        vector<ll> c(n);
        for (ll i = 0; i < n; ++i) {
            cin >> c[i];
        }
        ll s = accumulate(c.begin(), c.end(), 0LL);

        unordered_map<ll, ll> fact = factorized[s];
        for (ll i = 0; i < n; ++i) {
            for (const auto &[p, cnt] : factorized[c[i]]) {
                fact[p] -= cnt;
            }
        }

        ll ans = 1;
        for (const auto &[p, cnt] : fact) {
            ans = ans * atcoder::pow_mod(p, cnt, m) % m;
        }
        cout << ans << endl;
    }

    return 0;
}

// 1884ms

File: abc425/e/main2.cs
long PowMod(long x, long n, long m)
{
    if (!(x >= 0 && m >= 1)) throw new ArgumentOutOfRangeException();
    long res = 1;
    x %= m;
    while (n > 0)
    {
        if ((n & 1L) == 1L)
        {
            res = res * x % m;
        }
        x = x * x % m;
        n >>= 1;
    }
    return res;
}

string[] input0 = Console.ReadLine()!.Split(" ");
long t = long.Parse(input0[0]);
long m = long.Parse(input0[1]);

Span<bool> isPrime = new bool[5001];
for (int i = 0; i < isPrime.Length; ++i) isPrime[i] = true;
Span<int> minFactor = new int[5001];
for (int i = 0; i < minFactor.Length; ++i) minFactor[i] = -1;
isPrime[1] = false;
minFactor[1] = 1;
for (int p = 2; p < isPrime.Length; ++p)
{
    if (!isPrime[p]) continue;
    minFactor[p] = p;
    for (int q = 2 * p; q < isPrime.Length; q += p)
    {
        isPrime[q] = false;
        if (minFactor[q] == -1) minFactor[q] = p;
    }
}

Span<Dictionary<int, int>> factorized = new Dictionary<int, int>[5001];
factorized[1] = new();
for (int x = 2; x < factorized.Length; ++x)
{
    factorized[x] = new(factorized[x - 1]);
    int y = x;
    while (y > 1)
    {
        int p = minFactor[y];
        while (minFactor[y] == p)
        {
            y /= p;
            factorized[x][p] = factorized[x].GetValueOrDefault(p) + 1;
        }
    }
}

while (t-- > 0)
{
    int n = int.Parse(Console.ReadLine()!);
    int[] c = Console.ReadLine()!.Split(" ").Select(int.Parse).ToArray();
    int s = 0;
    foreach (int ci in c) s += ci;

    Dictionary<int, int> fact = new(factorized[s]);
    foreach (int ci in c)
    {
        foreach (var (p, cnt) in factorized[ci])
        {
            fact[p] -= cnt;
        }
    }

    long ans = 1;
    foreach (var (p, cnt) in fact)
    {
        ans = ans * PowMod(p, cnt, m) % m;
    }
    Console.WriteLine(ans);
}

// TLE(17/51)
